auth: new authentication library

deangereaux opened the pull request:
Origin: Original

New authentication library for authenticating devices, key features:
   - Two auth methods, DTLS and Challenge-Response.
   - Transport independent, initial support for serial and Bluetooth.
   - Multiple authentication instances.
   - Refer to ZAUTH RFC for more details:
      https://github.com/zephyrproject-rtos/zephyr/issues/23465

Signed-off-by: Dean Gereaux <deang@goldenbits.com>

deangereaux committed:
auth: new authentication library

Origin: Original

New authentication library for authenticating devices, key features:
   - Two auth methods, DTLS and Challenge-Response.
   - Transport independent, initial support for serial and Bluetooth.
   - Multiple authentication instances.
   - Refer to ZAUTH RFC for more details:
      https://github.com/zephyrproject-rtos/zephyr/issues/23465

Signed-off-by: Dean Gereaux <deang@goldenbits.com>

deangereaux commented on samples/authentication/serial/auth_server/sample.yaml:
```diff
@@ -0,0 +1,8 @@
+  name: Auth Server Serial
+tests:
+  sample.auth_server_serial:
+    harness: bluetooth
```
`bluetooth` harness probably isn't correct here, but without it I get a timeout error: `INFO    - 1/1 qemu_x86                  samples/authentication/serial/auth_server/sample.auth_server_serial FAILED Timeout (qemu)`

Any guidance on how to correctly use the harness option is appreciated.

deangereaux committed:
auth: Added missing copyright notice.

- Added correct copyright notice.

Signed-off-by: Dean Gereaux <deang@goldenbits.com>

d3zd3z reviewed (dismissed):
I'll start with a few higher-level questions that probably would have been better for the RFC review, but I seem to have missed that.

1. When using DTLS, it appears that this sets up and uses a DTLS connection to authenticate with the server. It then appears to no longer use that DTLS connection for upcoming traffic. This makes subsequent communication with this client fairly susceptible to interception. In other words, the client can easily authenticate a DTLS connection and both the server and client can know that connection is secure, but then if subsequent communication goes over another channel, that authentication was fairly pointless, since that subsequent communication can easily be bypassed. I guess in summary, I don't really see the point in authenticating with a DTLS connection but not actually using that connection for communication.
2. This applies even moreso for a challenge response. A simple challenge response can be useful to authenticate, when communication is going over an already secured channel (it is common for the client to authenticate with a server, once a TLS/DTLS connection has been established). Over an insecure channel, it suffers the same problems as above, even if the connection is authenticated, nothing stops subsequent unsecured communication from being intercepted or tampered with.

In other words, I don't see an arguable security benefit to this library as it is. Providing something to assist with establishing a DTLS connection could be useful, although we kind of already have that in the socket layer. Providing additional authentication also seems useful, but only as something that can be layered on top of an already secured channel, primarily to authenticate in the other direction.

In addition, I'd strongly suggest using an already established standard for a challenge-response, such as SCRAM that has address many of the security concerns around a more simplistic authentication.

d3zd3z commented:
There is definitely some merit in a library like this for communication over a reasonably secured channel, such as a physical connection to a device, or something where the underlying protocol uses authenticated encryption (such as BT). Using DTLS this way is still a little unusual, I think it would be better to have subsequent traffic go through the DTLS session. SASL does provide an X.509 authentication method that could be used for this type of authentication, but again, that is only really meaningful if the underlying channel is already secured in some manner.

deangereaux commented:
Thanks for reviewing and the feedback. 

A little history to provide some context.  This effort was stared from a requirement for a wireless lighting control system I worked on. The nodes and mobile application communicated over a secure BT link and had to be authenticated.  Since the light controls (small low cost, low power devices) needed to authenticate each other autonomously, a certificate based approach along with DTLS was ideal.  Especially since the secure hardware elements (Microchip ATECC608B) provided the ability to securely store X.509 certs and ECC keys.  Another project I worked on needed to authenticate a heart monitor chest pads, the pad output didn’t need to be encrypted.   

I do agree with you, not having the ability to encrypt subsequent messages over a link after authenticating can be of limited value.   I seriously contemplated adding encrypting subsequent messages, but wanted to limit the scope of this initial version.  However I’m happy to add this ability to make this authentication library more useful.  The authentication library design is pluggable, so adding SCRAM is very doable, I’ll investigate adding this.

If you want, we can talk more during the next API or Devel meeting, or some other time, just email me (deang@goldenbits.com).  Bottom line: I’m happy to work with the Zephyr team to make this Authentication library useful.

d3zd3z commented:
I guess it kind of comes down to what the goals are. There may be some use for a naive authentication over an unencrypted channel, when that channel could be considered secure, especially in very resource-constrained devices. As long as people using this realize that there isn't really very much security gained by doing this. Using an authentication mechanism over an already secure channel (such as a BT connection) is more useful.

Overall, this does seem like a reasonable approach.

I do wonder if something like GSSAPI would be too complex for the Zephyr environment. It has the advantage of being a standard API.

ceolin commented on samples/authentication/serial/auth_client/src/main.c:
```diff
@@ -0,0 +1,239 @@
+
+#if defined(CONFIG_AUTH_DTLS) && defined(CONFIG_AUTH_CHALLENGE_RESPONSE)
+#error Invalid authentication config, either DTLS or Challenge-Response, not both.
+#endif
```
This kind of logic should be in the build system. These options have to be mutually exclusive

deangereaux replied:
This is actually an error, removed code.

ceolin commented on lib/auth/auth_dtls.c:
```diff
@@ -0,0 +1,959 @@
+	mbedtls_ctr_drbg_init(&mbed_ctx->ctr_drbg);
+	mbedtls_entropy_init(&mbed_ctx->entropy);
+
+	sys_rand_get(mbed_ctx->cookie, sizeof(mbed_ctx->cookie));
```
don't you need cryptographically secure random numbers ? See `sys_csrand_get`

deangereaux replied:
Added new function `auth_get_random()` in auth_xport_common.c to handle SOCs with and without hardware based random number generators.

d3zd3z replied:
In general, this code probably shouldn't be supported on SOCs that don't have hardware based random numbers. I wouldn't go out of the way to make it work, since the end result won't be secure.

ceolin commented on lib/auth/auth_dtls.c:
```diff
@@ -0,0 +1,959 @@
+	if ((certs->device_cert_pem.priv_key == NULL) || (certs->device_cert_pem.priv_key_size == 0)) {
+		auth_free_mbedcontext(mbed_ctx);
+		LOG_ERR("Failed to get device private key");
+		return AUTH_ERROR_DTLS_INIT_FAILED;
```
this pattern is repetitive across this function, just log the problem and goto an error label

deangereaux replied:
Good point.  Added goto error label.

ceolin commented on lib/auth/Kconfig:
```diff
@@ -0,0 +1,77 @@
+		bool "Challenge Response (Must use TINYCRYPT and TINYCRYPT_SHA256)"
+		help
+		Selects the Challenge Response authentication method.
+		Must use TINYCRYPT and TINYCRYPT_SHA256
```
This dependency should be explicit.
```
depends on TINYCRYPT
depends on   TINYCRYPT_SHA256
```

deangereaux replied:
Fixed.

ceolin commented on lib/auth/CMakeLists.txt:
```diff
@@ -0,0 +1,24 @@
+if(CONFIG_MBEDTLS)
+    zephyr_include_directories($ENV{ZEPHYR_BASE}/../modules/crypto/mbedtls/include)
+    zephyr_include_directories($ENV{ZEPHYR_BASE}/../modules/crypto/mbedtls/configs)
+endif()
```
nope,  first I don't think you should check on `CONFIG_MBEDTLS`. Instead of this, `AUTH_DTLS` should depends on mbedtls and here your check should be something like:\
```
if (CONFIG_AUTH_DTLS)`
zephyr_library_link_libraries_ifdef(CONFIG_MBEDTLS mbedTLS)
endif()
```

the include path should already have been set to you.

deangereaux replied:
Thanks for the guidance here, fixed.

ceolin commented on lib/auth/Kconfig:
```diff
@@ -0,0 +1,77 @@
+		bool "DTLS Authentication"
+		help
+		Selects the DTLS authentication method.
+
```
something here, this code depends on mbedtls doesn't ?

deangereaux replied:
Fixed.

ceolin commented on lib/auth/auth_chalresp.c:
```diff
@@ -0,0 +1,637 @@
+{
+	int numbytes;
+
+	while ((int)msglen > 0) {
```
this is cast is wrong, if `msgelen` is smaller than `0` it means an overflow in `msgbuf`, if it is `0` you don't need to cast it to signed type. Inspect the code it does not seem  possible `msglen` be lesser than 0.

Another problem is that this cast can result in `< 0` for a big `msglen`.

deangereaux replied:
Fixed.

ceolin commented on lib/auth/auth_dtls.c:
```diff
@@ -0,0 +1,959 @@
+		break;
+	}
+
+	case 3:
```
use `__fallthrough` attribute in these cases.

deangereaux replied:
Fixed.

ceolin commented on lib/auth/auth_lib.c:
```diff
@@ -0,0 +1,342 @@
+{
+	/* Free any resources, nothing for now, but maybe
+	 * needed in the future */
+	return AUTH_SUCCESS;
```
don't we need to free any resource ? What happens with mbedTLS context ?

deangereaux replied:
Yes, that's an oversight.  Added code to properly de-init the Challenge-Response and DTLS auth methods.

ceolin commented on lib/auth/auth_lib.c:
```diff
@@ -0,0 +1,342 @@
+			}
+		}
+	}
+#endif
```
these `ifdef` don't scale well, if you have N backends this code will get very complicate. You should have an API that is implemented by the backend and here you should just call the init callback.

deangereaux replied:
Makes sense, the common code should not have to know about the individual auth method details. Fixed, moved backend specific init code to the backend init function.

ceolin commented:
I put some comments in parts of the code, but I share @d3zd3z concerns and I think we should elaborate / discuss more the pros/cons of adopting an existent standard

deangereaux committed:
auth: Fixed issues from review comments:

- Correct KConfig and CMakeLists.txt vars.
- Common function to get random numbers.
- Moved auth specific init into auth code vs. common auth_lib.c
- Addition misc code fixes and cleanup.

Signed-off-by: Dean Gereaux <deang@goldenbits.com>

deangereaux committed:
auth: Fixed issues from review comments:

- Correct KConfig and CMakeLists.txt vars.
- Common function to get random numbers.
- Moved auth specific init into auth code vs. common auth_lib.c
- Addition misc code fixes and cleanup.

Signed-off-by: Dean Gereaux <deang@goldenbits.com>

deangereaux commented on lib/auth/auth_chalresp.c:
```diff
@@ -100,32 +101,67 @@ struct auth_chalresp_result {
 
-/* default shared key if not set */
-static uint8_t *shared_key = default_shared_key;
+static struct chalresp_instance chalresp_inst_tbl[NUM_CHALLENGE_RESP_INST];
```
Found problem here were the Challenge-Response didn't correctly handle multiple instances.  Now it does.

deangereaux commented on lib/auth/auth_chalresp.c:
```diff
@@ -100,32 +101,67 @@ struct auth_chalresp_result {
+ * Challenge-Response instance info.  Used to set shared key for each
+ * auth instance.
  */
-static uint8_t default_shared_key[AUTH_SHARED_KEY_LEN] = {
```
Removed default shared key out of fear users of this library would just rely on this key.

deangereaux commented on lib/auth/auth_dtls.c:
```diff
@@ -622,189 +624,20 @@ static int auth_tls_entropy(void *data, unsigned char *output, size_t len,
- * @see auth_internal.h
- *
- */
-int auth_init_dtls_method(struct authenticate_conn *auth_conn, struct auth_dtls_certs *certs)
```
Not sure why git thinks this entire function was removed and re-added.

deangereaux commented on lib/auth/auth_xport_common.c:
```diff
@@ -1003,6 +1004,19 @@ void auth_message_hdr_to_be16(struct auth_message_frag_hdr *frag_hdr)
+/**
+ * @see auth_internal.h
+ */
+void auth_get_random(uint8_t *buf, size_t num)
```
New common Random function.

d3zd3z replied:
If we did agree to create a new random function, I think it would be appropriate to add it as its own change. However, I would argue that if there isn't a cryptographically secure random source available, the security added by this code is really more of an illusion. Non cryptographic random sources are rather predictable, and generally result in catastrophic weaknesses when used with cryptograph protocols, such as this.

dleach02 replied:
I don't think you should be defining yet another wrapper around rand. Decide whether your subsystem requires a cryptographically secure random values. If it does then just call sys_csrand_get() directly because that shows the reader the intent. If it doesn't then call sys_rand_get(). This way, if the underlying HW doesn't support a cryptographically secure random and someone tries to enable code that requires it, a build error will occur.

deangereaux committed:
auth: Removed incorrect config checks.

- Removed obsolete checks

Signed-off-by: Dean Gereaux <deang@goldenbits.com>

deangereaux committed:
auth: Removed incorrect config checks.

- Obsolete checks removed

Signed-off-by: Dean Gereaux <deang@goldenbits.com>

deangereaux commented:
@ceolin  Thanks for reviewing the PR.  Yes I agree we should discuss using a standard challenge-response.  Maybe during next Tuedays’ API meeting, if not Thurday's Devel call?  Hopefully @d3zd3z can join.

deangereaux committed:
auth: Added auth param for unit test.

- Added param for unit test

Signed-off-by: Dean Gereaux <deang@goldenbits.com>

deangereaux committed:
auth: Fixed formatting issues.

- Removed obsolete code comment.
- Fixed auth API doc file.
- Fixed sample proj conf files.

Signed-off-by: Dean Gereaux <deang@goldenbits.com>

deangereaux committed:
auth: Fixed formatting issues.

- Removed obsolete code comment.
- Fixed auth API doc file.
- Fixed sample proj conf files.

Signed-off-by: Dean Gereaux <deang@goldenbits.com>

deangereaux committed:
auth:  More formatting fixes.

- Removed trailing spaces in .rst doc file.
- Added tabs to samples.
- Fixed incorrect Kconfig define

Signed-off-by: Dean Gereaux <deang@goldenbits.com>

deangereaux committed:
auth:  More formatting fixes.

- Removed trailing spaces in .rst doc file.
- Added tabs to samples.
- Fixed incorrect Kconfig define

Signed-off-by: Dean Gereaux <deang@goldenbits.com>

deangereaux committed:
auth:  More formatting fixes.

- Fixed Kconfig in doc file.
- Added tabs to samples.
- Fixed incorrect Kconfig define

Signed-off-by: Dean Gereaux <deang@goldenbits.com>

deangereaux committed:
auth:  More formatting fixes.

- Fixed Kconfig in doc file.
- Added tabs to samples.
- Fixed incorrect Kconfig define

Signed-off-by: Dean Gereaux <deang@goldenbits.com>

deangereaux committed:
auth:  More formatting fixes.

- Fixed assignment in if() statement.
- Replaced spaces w/tabs in selected samples.

Signed-off-by: Dean Gereaux <deang@goldenbits.com>

deangereaux committed:
auth: Fixed incorrectly placed param check.

- Checking function parameters before function main body.

Signed-off-by: Dean Gereaux <deang@goldenbits.com>

deangereaux committed:
auth: Added ability to use DTLS after handshake

- An appplication can now send/recv messages via
  DTLS after a successful handshake.  Two new functions
  have been added: auth_lib_dtls_send()/recv().
- The Bluetooth samples have been updated to echo
  DTLS messages using these two new functions.

Signed-off-by: Dean Gereaux <deang@goldenbits.com>

deangereaux commented:
@d3zd3z  I added new functions to send/recv messages using DLTS after the initial handshake.  This provides some nice additional functionality for users.   For authentication methods besides DTLS,  a “hook” can be added enabling the user to write their own encrypt/decrypt code.  I’d like to get some user feedback and/or requests for this functionality before adding it.  

Regarding GSSAPI,  this might be something that’s layered on top of Zauth eventually.   One of my goals was to keep the interface simple to encourage use vs. a heavier API like GSSAPI.

dleach02 reviewed (changes requested):
You need to clarify the subsystem requirements for rand()

deangereaux committed:
auth: Requiring use of cryptographically secure rand generator.

- Updated code to use sys_csrand_get().
- Updated docs to indicate secure random generator is required.
- Added explicit doc note warning the Challenge-Response
  auth method does not encrypt subsequent messages of the link.
- Also fixed varous formatting errors.

Signed-off-by: Dean Gereaux <deang@goldenbits.com>

deangereaux committed:
auth: Updated CODEOWNERS

- Added @deangereaux to /lib/auth/ and /samples/authentication

Signed-off-by: Dean Gereaux <deang@goldenbits.com>

deangereaux committed:
auth: Fixed formatting issues.

- Added correct spacing for if() and switch() statements.

Signed-off-by: Dean Gereaux <deang@goldenbits.com>

deangereaux commented:
@dleach02  Latest commit removes wrapper around  *_rand() and explicitly uses sys_csrand_get()

deangereaux commented:
@ceolin @d3zd3z   I added an explicit warning in the docs about how the Challenge-Response does not provide security for subsequent messages and how DTLS can be used for follow-on secure messages.

I spent some time reviewing/read specs for GSSAPI.   It seems GSSAPI is suited for handling authentication for web base apps or enterprise level networks vs. small embedded devices which ZAUTH is targeting.  Regarding SCRAM, mentioned in this PR, might be good to add later.  All of the Challenge-Response uses cases I’ve worked on are straight forward and do not require something like SCRAM.  For example, authenticating a printer cartridge or an attached display.  But yes this all depends on the application and risk associated with exposing printer commands or displaying incorrect messages.   

Regardless, I’m happy to talk and decide what is the best path forward for this PR.
Thanks

d3zd3z commented:
I do think there are decent uses of a simplistic authentication library for Zephyr, as long as we are clear that there are security tradeoffs being made with regards to code size and complexity. For something like authenticating an attached piece of hardware, and the likes, just authenticating, and then not worrying about the security of the resulting transmission probably makes sense.

deangereaux committed:
auth: Added CONFIG_ENTROPY_HAS_DRIVER to test cases.

- Fixes problem with test case not compling.

Signed-off-by: Dean Gereaux <deang@goldenbits.com>

deangereaux committed:
auth: Fixed problem with running unit test

- Added back sys_rand_get() for CI tests.  This should not be used for
  production code, but needed for CI tests.
- Updated serial sample YAML files to used ci.prf.conf file for
  CI purposes.

Signed-off-by: Dean Gereaux <deang@goldenbits.com>

deangereaux commented:
@dleach02    I had to add back the insecure random number generator, `sys_rand_get()`, in order to pass CI tests. Maybe there’s a way to use `sys_csrand_get()` when running twister tests, but dang I couldn’t find it.  I did add warnings in the code,  I’ll also update the documentation warning about using the insecure generator.

d3zd3z commented:
> @dleach02 I had to add back the insecure random number generator, sys_rand_get(), in order to pass CI tests.

This is a bit deeper of a problem than just this change. At one point, we had an insecure source of entropy available, but this is a pretty bad idea if it makes it into an actual device. (e.g. [Netscape](http://people.eecs.berkeley.edu/~daw/papers/ddj-netscape.html)).

@dleach02 How do we solve this for other users of the crypto random sources? Do we need to come up with a more general solution here? It is probably ok to have insecure random numbers on CI targets, but we really need to make sure that never makes it onto an actual device.

I don't think the solution is to put the weakness into this auth code. The auth code needs strong random numbers, and so should be asking for them. We just need to figure out how to test this in CI.

deangereaux committed:
auth: Removed non-secure random generator, added stub func for CI tests

- Removed sys_rand_get(), must use secure sys_csrand_get()
- Added stub function, z_impl_sys_csrand_get() in test code.
  This enables unit testing without test environment having
  a secure random source.
- Requring samples to be build with nrf52840dk_nrf52840 board.
  Updated sample.yaml files

Signed-off-by: Dean Gereaux <deang@goldenbits.com>

deangereaux committed:
auth: Formatting fix

- Replaced leading spaces with tab

Signed-off-by: Dean Gereaux <deang@goldenbits.com>

deangereaux commented:
@d3zd3z  I look hard at the test tools and system and didn’t find any way to mock out the sys_csrand_get() call.   As a solution I added a sub function in the test code here:  test/lib/authentication/sr/main.c.

deangereaux commented on tests/lib/authentication/src/main.c:
```diff
@@ -83,3 +83,20 @@ void test_main(void)
+ *
+ * @return  0 on success.
+ */
+int z_impl_sys_csrand_get(void *dst, size_t len)
```
@d3zd3z   Here's the stub function just for CI testing.

deangereaux commented on samples/authentication/bluetooth/central_auth/sample.yaml:
```diff
@@ -3,6 +3,7 @@ sample:
+    build_only: true
     harness: bluetooth
-    platform_allow:  qemu_x86
+    platform_allow: nrf52840dk_nrf52840
```
For CI, I added the nrf52840dk_nrf52840 board as the only allowed platform.  This is successful if you run twister with --all option like this:  `twister --all samples/authentication/bluetooth/central_auth/sample.central_auth`

d3zd3z commented on tests/lib/authentication/src/main.c:
```diff
@@ -0,0 +1,102 @@
+int z_impl_sys_csrand_get(void *dst, size_t len)
+{
+    /* Not random, for testing purposes only. */
+    memset(dst, 42u, len);
```
42 is a reasonably good choice, although https://xkcd.com/221/ 4 is also common :-)

deangereaux committed:
auth: Formatting fix

- Fixed various formatting issues.

Signed-off-by: Dean Gereaux <deang@goldenbits.com>

deangereaux committed:
auth: Formatting fix

- Fixed various formatting issues.

Signed-off-by: Dean Gereaux <deang@goldenbits.com>

deangereaux committed:
auth: Merge fix, incorrect path

- Fixed incorrect document path

Signed-off-by: Dean Gereaux <deang@goldenbits.com>

deangereaux committed:
auth: Updated sample prj.conf files.

- Removed BT_PRIVACY to simplify config

Signed-off-by: Dean Gereaux <deang@goldenbits.com>

deangereaux committed:
auth: Removed unnecessary typedef

- Typedef not needed for table.

Signed-off-by: Dean Gereaux <deang@goldenbits.com>

deangereaux committed:
auth: Minor formatting fix.

- Removed extra space.

Signed-off-by: Dean Gereaux <deang@goldenbits.com>

deangereaux commented:
There’s only one compliance check that is failing, a typedef (in auth_xport.h,  `typedef void *auth_xport_hdl_t`) used to define an intentionally opaque transport handle.  This should be OK per the Linux coding standards

deangereaux commented:
@carlescufi @d3zd3z @dleach02 @galak @MaureenHelm @nashif @ceolin @frkv   This ZAUTH PR has been aging and like a good wine it is getting better and better!!   I’d like to push this PR to the finish line here and merge.  This library will be marked as EXPERIMENTAL, as such any issues can be readily addressed reducing any risk. 
Can you help me by reviewing and approving this PR?  
Thanks, Dean

deangereaux commented:
There doesn’t seem to be any interest in going forward and merging this ZAUTH PR.   Maybe my perception is incorrect, however I’m not seeing any feedback or comments.  I do think ZAUTH is useful and provides a solution for authentication use cases I personally have had to write code to solve.  However, if the larger Zephyr community is not interested in ZAUTH, then let’s close this PR and put it on the shelf.

deangereaux commented:
Closing.
